File: leetcode/bt_merge.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def mergeTrees(self, t1, t2):
        """
        :type t1: TreeNode
        :type t2: TreeNode
        :rtype: TreeNode
        """
        self.mergeNode(t1, t2)
        return t1

    def mergeNode(self, n1, n2):
        # 因为要merge到n1，一定要保证n1有子节点先
        if n1 and not n1.left and n2 and n2.left:
            n1.left = TreeNode(0)
        if n1 and not n1.right and n2 and n2.right:
            n1.right = TreeNode(0)

        if n1 and n2:
            n1.val += n2.val
            self.mergeNode(n1.left, n2.left)
            self.mergeNode(n1.right, n2.right)
            return

        if n1:
            self.mergeNode(n1.left, None)
            self.mergeNode(n1.right, None)
            return

        if n2:
            n1 = TreeNode(n2.val)
            n1.left = TreeNode(n2.left.val)
            n1.right = TreeNode(n2.right.val)

            self.mergeNode(n1.left, n2.left)
            self.mergeNode(n1.right, n2.right)
            return

        return

class Solution2():

    def mergeTrees(self, t1, t2):
        dst = TreeNode(0)
        return self.mergeNode(dst, t1, t2)

    def mergeNode(dst, n1, n2):
        if n1:
            dst.val += n1.val

        if n2:
            dst.val += n2.val

        if n1 and n1.left or n2 and n2.left:
            dst.left = TreeNode(0)
            l1 = None
            l2 = None
            if n1:
                l1 = n1.left
            if n2:
                l2 = n2.left
            self.mergeNode(dst.left, l1, l2)

        if n1 and n1.right or n2 and n2.right:
            dst.right = TreeNode(0)
            r1 = None
            r2 = None
            if n1:
                r1 = n1.right
            if n2:
                r2 = n2.right
            self.mergeNode(dst.right, r1, r2)

File: leetcode/bt_merge.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

struct TreeNode {
        int val;
        struct TreeNode *l;
        struct TreeNode *r;
};

typedef struct TreeNode Node_t;

// Node_t *merge(Node_t *a, Node_t *b)
// {
//         Node_t *dst = malloc(sizeof(Node_t));
//         if (!a && !b)
//                 return dst;
//         dst->val = a ? a->val : 0;
//         dst->val += b ? b->val : 0;
//         merge(dst->left, merge(a->left, b->left));
//         merge(dst->right, merge(a->right, b->right));
//         return dst;
// }
//
// struct TreeNode* mergeTrees(struct TreeNode* t1, struct TreeNode* t2) {
//         struct TreeNode *dst = malloc(sizeof(struct TreeNode));
//         if (!t1 && !t2)
//                 return dst;
//         dst->val = t1 ? t1->val : 0;
//         dst->val += t2 ? t2->val : 0;
//
//         mergeTrees(dst->left, mergeTrees(t1 ? t1->left : 0, t2 ? t2->left : 0));
//         mergeTrees(dst->right, mergeTrees(t1 ? t1->right : 0, t2 ? t2->right : 0));
//         return dst;
// }

static int count = 0;
Node_t *tdata(int *vals, int s, int *c)
{
        Node_t *root;
        if (*c <= s) {
                root = calloc(sizeof(Node_t), 1);
                root->val = *(vals + *c);
                printf("val %d\n", *(vals + *c));
                *c = *c + 1;
        }
        else
                return NULL;

        if (count++ % 2)
                goto RIGHT;
LEFT:
        if (s > 1) {
                s--;
                printf("c: %d left %d\n", *c);
                root->l = tdata(vals, s, c);
                // goto RIGHT;
        }
RIGHT:
        if (s > 2) {
                s--;
                printf("c: %d right %d\n", *c, *(vals + *c));
                root->r = tdata(vals, s, c);
        }
        return root;
}


int main(int argc, char **argv)
{
        int tvals_a[8] = {1, 3, 2, 5, 3, 6, 2, 6};
        int s = 8;
        int *c = malloc(sizeof(int));
        *c = 0;
        Node_t *ta = tdata(tvals_a, s, c);
        return 0;
}
